### ISSUE TITLE
Add support for Easee car charger and generic charging devices

### ISSUE THREAD

**Issue by frodeheg** [frodeheg/no.sparegris#4]
Requirements:
* There will be support for only 1 charger (Is this limitation bad? Should I make it 2?)
* Chargers with adjustable currents must be able to adjust it up/down
* The charger should not switch on/off a lot.
* The priority of the charger should be set somehow.

Assumptions:
* Setting a higher "free power" than the lowest charging rate as a trigger for starting charging will eliminate excessive switching off as the charger will not use up all the free power when being turned on.
* Charging always starts at a minimum charging current and is increased from there (as fast as the throttle time allows up to the monthly power tariff)
* The monthly power tariff can be considered as a maximum charging power

* In the app you can configure:
  * A trigger when to start charging ( > x watts available)
  * The minimum charging power in W (or can I say that this is 75% of the trigger and just ignore having a setting for it?)
  * A throttle count how often to send signals to the charger (going from 5 seconds to 5 minutes) - will be overridden by emergency off though.
* App receives signals
  * Start charging (with end time and minimum kWh to be offered to the charger within this time - The offered power will be distributed on the lowest cost hours between now and the end time. Rescheduling will be recalculated every new hour)
* App sens out signals (only if a start charging signal has been received)
  * Turn on charger (based on a trigger > x watts available ; assumes charging starts at the lowest level)
  * Increase charging current (You may choose to ignore this signal or its parameter; parameter being "Maximum increase in W")
  * Decrease charging current (You may choose to ignore this signal or its parameter; parameter being "Minimum decrease in W")
  * Turn off charger (only as a last resort if decrease charging current was not sufficient)

**Option 1: Do the controls entirely with flows:**
* There will be no internal car charger control, e.g. The charger should not be a controllable device.
* Flows must be created using the actions / triggers as listed above
* Use the free power threshold already in the app to set the priority
* Assumes that charging starts at the minimum charging power

**Option 2: Do the controls entirely within the app**
* The car charger device must be controlled by the app.
* A flow should still be used to start the charging (to indicate minimum kWh to charge by which time)
* Priority set by device listing

**Option 3: Both of the above possible, with a selector**
* Can only choose one of the options

**Comment by frodeheg** [comment]
Support added in https://github.com/frodeheg/no.sparegris/commit/109e7f678383f25cb1e6457f2fa5544cd6afd67b

**Comment by dooniem** [comment]
TLDR; Keep both "trigger" and "minimum", multiply the trigger by 25% on "R months" because of ovens and allow as many chargers as the user pleases.

> There will be support for only 1 charger (Is this limitation bad? Should I make it 2?)

As far my understanding go, I believe you can allow/support as many chargers as the user pleases – at least when talking about Easee chargers. The developer of the Easee integration mentioned that "set dynamic circuit current" would affect all connected chargers. This is similar to a how commercial chargers operates, all available current is evenly distributed. If the user want's to prioritize, then he would have to set up extra flows to pause the one(s) not prioritized. Also be aware that even if 20A is offered, each car will only draw what they are able to draw, meaning min and max effects is totally "fuse based", not car or number of charger based. But as the developer, you might see other problems offering 2 or more?

> Charging always starts at a minimum charging current and is increased from there

I wonder if the app will increase in sequential Ampere-steps, or calculate based on remaining Wh. If it's the first, setting 5 minutes throttle time and allowing a low charging power could end up in never reaching the max drain even when there's available power. Assuming it reverts back to minimum at the start of a new hour.

> The minimum charging power in W (or can I say that this is 75% of the trigger and just ignore having a setting for it?)

My first thought is that trigger power should vary with season. During winter I would want a higher trigger.
I guess that having a set number and add 25% to that number in months containing an "R" would be sufficient.
My second thought is that the trigger could be calculated based on minimum charge, ie. the other way around of what you're suggesting. But if you choose this 25% will be too little. Say you set 1500W minimum, then the trigger would be 1875W.
Having Easse only accept integers, 6,5A would send 7A leaving you with 14% wiggle room instead of your intended 25% – assuming 230V. 

As I know you know, Easee's recommendation is to never go below 6A (7A for Renault Zoe) as some cars will not handle it. I have no idea, but **I assume** different cars will handle this differently, maybe one brand will have the power go to waste, maybe one brand will report an error having the user re-plug the car before able to continue. Meaning the charge session will fail from the get go. Again, just speculation. I've experienced re-plugging on "fast chargers".  My 5 cents is to either inform on this matter, or make it impossible to set it lower. Whatever you choose, my aim will be to pause/stop charging when the charger reports below 7A. There's always a new hour ahead with more available power.

**Comment by dooniem** [comment]
Could you make a bullet point on how price will affect PBs strategy? I reckon if a user plugs in during daytime, needs power for a late night drive and sets the time accordingly... I just want to know that PBs initial plan (before high consumption in the household ruins it) will be to offer less Watts on higher priced hours than on low prices?

**Comment by frodeheg** [comment]
> But as the developer, you might see other problems offering 2 or more?

I'm already a bit exhausted thinking about all the hassle, so I wanted to defer the problem for later.

> I wonder if the app will increase in sequential Ampere-steps

Yes, if done directly. If by flow it would be the responsibility of the flow creator.

>  ...setting 5 minutes throttle time...

I think this will be too high, I will default it to 30 seconds or something like that.

> My first thought is that trigger power should vary with season.

If it's per season, then maybe it's ok to change it manually twice a year? ..like you have to for the delta-temperatures on heat pumps. (for now) Seasonal control lies a bit into the future I think. I don't think I can turn the heat-pump automatically on a given date every year for example, as it depends on the weather.

> My 5 cents is to either inform on this matter, 

I suppose this can be read from the charger app directly so I'll leave it to the users themselves to create informative flows.

> ... or make it impossible to set it lower.

For flow users I'll leave it to them. For app internal, I can set minimum to 7 A. ... 

> Could you make a bullet point on how price will affect PBs strategy?

Added: The offered power will be distributed on the lowest cost hours between now and the end time. Rescheduling will be recalculated every new hour

**Comment by frodeheg** [comment]
> power should vary with season.

Actually I see no problems with this. You can set it to the "winter" value all year. Only difference is that you will start charging a bit earlier in the summer.... It will still be charged just as much as it will try to use up the budget within the hour it is activated in.

**Comment by dooniem** [comment]
> Winter

Hmm. I’m totally open to being wrong on my assumption. I interpret the described trigger as a replacement of today’s “margin”. Meaning that the trigger will not only control the start, but in effect also control the end goal / aim in Wh before turning off the charger. I.e tariff 5000Wh, trigger 2000W and min charge 1500W would turn off the charger when power usage reaches 4500Wh. Or am I wrong already? 

If we can call it “PBs car charger Wh aim”, my line of thought is that bigger slack is needed during winter than summer to account for controlled and uncontrolled devices. 
I rather there be room for 2-3 ovens, controlled by PB or not, to heat my kid’s room during the night, rather than letting it cool for 5-10 mins before a new hour start. Not sure if that is a realistic case or not? Suddenly the hot water tank, dishwasher starts drying and my Quooker water tap starts heating, totaling way more Watts than the minimum charge from the charger. None of them controlled by PB. Since my actual charge max is lower than my tariff, and PB wants to get as close to tariff as possible, it might run a more risky game than it would have done if my charge max was 22kW, being able to boost at the end.

Does any of this makes sense?

**Comment by frodeheg** [comment]
> Meaning that the trigger will not only control the start

The idea was that it will only control when to start.

There are already two methods to ensure we do not overshoot, I think those should be sufficient.
Reserved power (which defaults to 500W) and error margin in %

> ... rather than letting it cool for 5-10 mins before a new hour start. Not sure if that is a realistic case or not?

If the priority of the charger is lower than the heaters then PB would try to reduce the power of the charger first. If the response time of the charger is slower than the next power change from the energy meter then the devices with higher priority will go off yes, but only for 1 minute if it's close to a new hour (5 minutes if it's in the beginning of the hour though). That being said, you can always put some other devices like water heater in between to reduce the likelihood of those specific heaters going off.

> totaling way more Watts than the minimum charge from the charger. None of them controlled by PB.

That is why you should set the trigger (which only acts as a trigger) higher, which will delay the start of the charger until the end of the hour when the charger can afford to reduce the power by as much as the uncontrollable devices use alone.

That being said, the charging does not stop when a new hour is entered. It will continue charging until it has to be turned off... then it will have to wait for a new trigger event to be turned on again. Hopefully this will only turn the charging off/on at most one time every hour and in many cases it will not be turned off at all during the hour keeping a more stable charging current.

Finally, when converting available power to Amps I will round down in order to prevent the charging current to go up/down all the time. Rather it may go up a little bit in the end of the hour.

**Comment by frodeheg** [comment]
Added support for all of the above except flow cards in this check-in: https://github.com/frodeheg/no.sparegris/commit/aee4bb9787eb25ba9ef93bb2a8b80e48796cdc16

**Comment by dooniem** [comment]
Sorry you have to spoon-feed me. 
I'm starting to think that what I'm asking is more or less controlled by the error margin. The less devices controlled by PB, the higher the uncertainty of hitting PB's targeted tariff. Right now I only have my car charger controlled by PB, meaning I have a lot of power heavy devices that can make life hard for PB.  Am I assuming correctly if "reserved 500W" and 5% margin means PB "strive for (5000W-500W)-5% = 4275Wh. In other words, if **all** electrical devices inside a house was controlled by PB including a car charger. Each hour would say 4275Wh on elhub? Or how should I interpret this?

If correct, maybe even the error margin should be season dependent and adjusted each time you smartify a device? :p 
Have you considered inputting the margin in Watts? Since much else is in Watts and probably easier for each household to estimate given their current setup? Though I do understand that % will be tariff independent...It's just harder for my brain to guesstimate a buffer/margin needed in percentages. 
I understand that the helptexts for how all this fits together, how the user best can adjust the default settongs for his current setup, is not priortized at the moment. But that's why all these basic questions pops up O:)

**Comment by frodeheg** [comment]
> Your first draft for car charging had "reserve power" as a term, I might have gotten your terms confused as well.

Yeah, I got a headache thinking of how to do this. It was so much back and forth that I had to write down some kind of specification here to get clarity into what my goal was too. I have written and deleted so much code because I didn't do the specification beforehand.

> Am I assuming correctly if "reserved 500W" and 5% margin means PB "strive for (5000W-500W)-5% = 4275Wh

The nice thing about reserving W is that it should not affect the limit in Wh (given that you're able to use the extra power at the end of the hour). So if you said the limit was 5 kWh with 5% error margin, then the limit will be 4.75 kWh regardless of how much power you reserve, so regardless of the season. (although In the current upcoming code release the maximum charging  power for the car will be the same as the hourly limit, so it might not be able to eat up that extra 250Wh in the end.)

The code to control the above is complete but I have no triggers and actions written yet so I will send out a new test version once that is complete.

**Comment by dooniem** [comment]
Ok, so back to my main concern, the uncontrolled loads that might start/fire at the end of an hour. I basically control this with the percentage only. I’m not able to grasp the thoroughness of your code and how this plays out when PB can’t control my 5 ovens, my 300L water heater, my 2 heat pumps, my 3 floor heatings, my pre-warm element in the balanced ventilation, and my 3L Quooker tank which will fire at will.
Would you say in theory I could experience overshooting 5kWh and have to adjust the percentage accordingly? If so I’d feel inputting Watts instead of percentage would be more like the rest of your settings and easier, since I’d have to calculate what the percentage represent anyway?

**Comment by frodeheg** [comment]
> Maybe it’s painfully obvious that the percentage should reflect your home’s uncontrolled loads.

Hehe, no, it's the opposite. The percentage should NOT be used to reflect the uncontrolled loads. The percentage is intended to only reflect inaccuracies in the estimated usage due to timing differences between Homey and the meter reader. Although, you might use it to artificially reduce the limit more, but that will not help you with the problematic case because it does not reserve power for uncontrolled devices.

If you have uncontrolled loads of 4000W and you're afraid all of them will  trigger at the end of every hour then you can reserve 4000W. However this is not necessary at all as long as the combined load for all uncontrollable devices is below the hourly threshold.

In the current version reserved power is set aside on top of the controlled devices. This is because not all controlled devices report power, however I have been speculating on including all controlled devices that report power in the reserved power because they all can contribute towards keeping us within the limit by turning them off when x amount of uncontrolled power suddenly is used. This would allow them to stay on for the full hour even if you reserve immense amounts.

With car chargers I am inclined to let the car-charger be included in the reserved power because those

**Comment by frodeheg** [comment]
For the action card "Start charging cycle" what is the best parameter?
* End time

or
* Duration

In case of end time this will be the first occurrence of the given time. (must be specified in hours:minutes)
In case of duration, it will be the number of minutes/seconds (apparently I cannot choose hours, but the strangest thing is that the duration property has as a special feature: if you use a dash you will get a random duration between two specified durations such as 2-10 would be a random duration between 2-10 minutes, not that it make sense for charging, but....)

If nobody complain I will only make the one with "end time".

**Comment by frodeheg** [comment]
Planning to charge is blocked by issue #12

**Comment by frodeheg** [comment]
For the planning I will put a default assumption that 75% power will be available for the charger. Once charging has been performed the number will be refined to improve future planning. There will be a separate number for every hour of the day. Of course, the number will also heavily depend on price point, but I don't find it necessary to delve into such details.
